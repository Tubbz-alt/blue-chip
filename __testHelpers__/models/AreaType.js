import BaseModel from "../../src/BaseModel";
import Spec from "./Spec";

export default class AreaType extends BaseModel {}
