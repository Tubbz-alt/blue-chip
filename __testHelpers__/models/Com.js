import BaseModel from "../../src/BaseModel";
import SpecDetail from "./SpecDetail";

export default class Com extends BaseModel {
  static get belongsTo() {
    return [SpecDetail];
  }
}
