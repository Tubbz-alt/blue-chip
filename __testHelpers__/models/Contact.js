import BaseModel from "../../src/BaseModel";
import VendorContact from "./VendorContact";

export default class Contact extends BaseModel {
  static get belongsTo() {
    return [VendorContact];
  }
}
