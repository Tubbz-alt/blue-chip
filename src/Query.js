import {isFunction} from "./utils";
export default class Query {
  constructor(klass, resourceName, resources, hasMany = [], belongsTo = []) {
    this.klass = klass;
    this.resourceName = resourceName;
    this.resources = resources;
    this.currentIncludes = [];
    this.currentResources = {};
    this.hasMany = hasMany;
    this.belongsTo = belongsTo;
    this._setCurrentResources();
  }

  find(id) {
    const {
      resources,
      resourceName,
      klass,
      _convertToModel,
      hasMany,
      belongsTo
    } = this;
    const {attributes} = resources[resourceName] && resources[resourceName][id];
    return _convertToModel(
      klass,
      resources,
      {id, ...attributes},
      hasMany,
      belongsTo
    );
  }

  first() {
    const {resources, resourceName} = this;
    const _resources = resources[resourceName];
    const _index = resources.index[resourceName];
    return _resources && _index && _resources[_index[0]];
  }

  last() {
    const {resources, resourceName} = this;
    const _resources = resources[resourceName];
    const _index = resources.index[resourceName];
    return _resources && _index && _resources[_index[_index.length - 1]];
  }

  all() {
    return this;
  }

  where(params) {
    this._filterAndSetCurrentResourcesByParams(params);
    return this;
  }

  whereRelated(relationship, params) {
    const {resourceName} = this;
    this.hasMany.includes(relationship)
      ? this._handleHasManyWhereRelated(relationship, params, resourceName)
      : this._handleBelongsToWhereRelated(relationship, params, resourceName);
    return this;
  }

  _handleBelongsToWhereRelated(relationship, params, resourceName) {
    const relationshipName = relationship.singularName();

    const relationIds = this.klass
      .query(this.resources)
      .includes([relationshipName])
      .toModels()
      .reduce((idArray, model) => {
        const maybeRelation = model[relationshipName];
        const relation = isFunction(relation) ? maybeRelation() : maybeRelation;

        if (relation && relation.id && !idArray.includes(relation.id))
          idArray.push(relation.id);
        return idArray;
      }, []);

    const filteredRelationIds = relationship
      .query(this.resources)
      .includes([resourceName])
      .where({id: relationIds})
      .where(params)
      .toObjects()
      .map(r => r.id);

    this.currentResources = Object.entries(this.currentResources).reduce(
      (newResources, [id, resource]) => {
        const r = resource.relationships[relationshipName];
        if (r && filteredRelationIds.includes(r.data.id)) {
          newResources[id] = resource;
        }
        return newResources;
      },
      {}
    );
  }

  _handleHasManyWhereRelated(relationship, params, resourceName) {
    this.currentResources = relationship
      .query(this.resources)
      .where(params)
      .includes([resourceName])
      .toObjects()
      .reduce((newResource, relatedResource) => {
        const relation =
          relatedResource[resourceName] ||
          [relatedResource[this.klass.singularName()]].filter(Boolean);
        relation.forEach(({type, id, ...attributes}) => {
          newResource[id] = {type, id, attributes};
        });
        return newResource;
      }, {});
  }

  includes(relationshipTypes) {
    this.currentIncludes = relationshipTypes;
    return this;
  }

  toModels() {
    if (!this.currentResources) return [];
    return this._reduceCurrentResources("models");
  }

  toObjects() {
    if (!this.currentResources) return [];
    return this._reduceCurrentResources("objects");
  }

  // Private

  _sortByIndex(resource1, resource2, resources, resourceName) {
    const index = resources.index[resourceName];
    return index.indexOf(resource1.id) - index.indexOf(resource2.id);
  }

  _reduceCurrentResources(reducerType) {
    // TODO: needs to be refactored
    const conversion =
      reducerType === "models" ? this._convertToModel : this._convertToObject;
    const {
      currentIncludes,
      currentResources,
      resources,
      resourceName,
      _flattenRelationships,
      hasMany,
      belongsTo
    } = this;

    return Object.values(currentResources)
      .sort((resource1, resource2) =>
        this._sortByIndex(resource1, resource2, resources, resourceName)
      )
      .map(({id, attributes, relationships, types, links}) => {
        const newFormattedResource = conversion(
          this.klass,
          resources,
          {
            id,
            ...attributes
          },
          hasMany,
          belongsTo
        );

        if (!currentIncludes.length) return newFormattedResource;
        return conversion(
          this.klass,
          resources,
          {
            ...newFormattedResource,
            ..._flattenRelationships(relationships).reduce(
              (nextRelationshipObjects, {id, name, type}) => {
                let relationClass = this.hasMany.find(klass => {
                  return klass.pluralName() === type;
                });

                if (relationClass) {
                  return this._handleHasManyIncludes(
                    resources,
                    id,
                    type,
                    nextRelationshipObjects,
                    conversion,
                    relationClass,
                    currentIncludes,
                    name
                  );
                }

                relationClass = this.belongsTo.find(klass => {
                  return klass.pluralName() === type;
                });

                if (relationClass) {
                  return this._handleBelongsToIncludes(
                    resources,
                    id,
                    type,
                    nextRelationshipObjects,
                    conversion,
                    relationClass,
                    currentIncludes,
                    name
                  );
                }

                return nextRelationshipObjects;
              },
              {}
            )
          },
          hasMany,
          belongsTo
        );
      });
  }

  _handleHasManyIncludes(
    resources,
    id,
    type,
    nextRelationshipObjects,
    conversion,
    relationClass,
    currentIncludes,
    name
  ) {
    const directIncludesRalationships = currentIncludes.map(
      relation => relation.split(".")[0]
    );
    if (!directIncludesRalationships.includes(name))
      return nextRelationshipObjects;
    if (!(name in nextRelationshipObjects)) {
      nextRelationshipObjects[name] = [];
    }
    if (!resources[type]) return nextRelationshipObjects;
    const relationData = resources[type][id];
    if (!relationData) return nextRelationshipObjects;

    const nestedResourceName = currentIncludes
      .filter(relation => relation.split(".")[0] == name)[0]
      .split(".")[1];

    if (relationClass) {
      let relationModel;
      if (nestedResourceName) {
        let nestedClass = relationClass.belongsTo.filter(
          klass => nestedResourceName === klass.singularName()
        )[0];

        if (nestedClass) {
          relationModel = this._convertToModel(
            relationClass,
            resources,
            {
              id,
              ...relationData.attributes
            },
            relationClass.hasMany,
            relationClass.belongsTo
          );
        }
      }

      nextRelationshipObjects[name].push(
        conversion(relationClass, resources, {
          id,
          ...relationData.attributes,
          ...(relationModel &&
            relationModel[nestedResourceName]() && {
              [nestedResourceName]: relationModel[
                nestedResourceName
              ]().toObject()
            })
        })
      );
    }
    return nextRelationshipObjects;
  }

  _handleBelongsToIncludes(
    resources,
    id,
    type,
    nextRelationshipObjects,
    conversion,
    relationClass,
    currentIncludes,
    name
  ) {
    if (!currentIncludes.includes(name)) return nextRelationshipObjects;

    if (!(name in nextRelationshipObjects)) {
      nextRelationshipObjects[name] = null;
    }

    if (!resources[type]) return nextRelationshipObjects;
    const relationData = resources[type][id];
    if (!relationData) return nextRelationshipObjects;

    if (relationClass) {
      nextRelationshipObjects[name] = conversion(relationClass, resources, {
        id,
        ...relationData.attributes
      });
    }

    return nextRelationshipObjects;
  }

  _convertToModel(klass, resources, resource, hasMany, belongsTo) {
    return new klass(resources, resource, hasMany, belongsTo);
  }

  _convertToObject(klass, resources, resource, hasMany, belongsTo) {
    return resource;
  }

  _flattenRelationships(relationships) {
    if (!relationships) {
      return [];
    }

    return Object.entries(relationships).reduce(
      (nextRelationships, [name, relationshipItem]) => {
        if (!nextRelationships || !relationshipItem || !relationshipItem.data) {
          return nextRelationships;
        }

        if (Array.isArray(relationshipItem.data)) {
          const dataArray = relationshipItem.data.map(item => ({
            ...item,
            name
          }));
          return [...nextRelationships, ...dataArray];
        }

        return [...nextRelationships, {...relationshipItem.data, name}];
      },
      []
    );
  }

  _setCurrentResources() {
    if (this._isEmpty(this.currentResources) && this.resources) {
      this.currentResources = this.resources[this.resourceName];
    }
  }

  _filterAndSetCurrentResourcesByParams(params) {
    if (!this.currentResources) return;
    const resourcesByID = Object.entries(this.currentResources).reduce(
      (newResource, [id, resource]) => {
        this._filterResourceByParams(params, newResource, resource, id);
        return newResource;
      },
      {}
    );
    this.currentResources = resourcesByID;
  }

  _filterResourceByParams(params, newResource, resource, id) {
    Object.entries(params).forEach(([key, value]) => {
      if (Array.isArray(value)) {
        if (key === "id" && value.includes(resource.id)) {
          newResource[id] = resource;
        } else if (value.includes(resource.attributes[key])) {
          newResource[id] = resource;
        }
      } else {
        if (key === "id" && resource.id === value) {
          newResource[id] = resource;
        } else if (resource.attributes[key] === value) {
          newResource[id] = resource;
        }
      }
    });
  }

  _isEmpty(obj) {
    if (
      obj === null ||
      obj === undefined ||
      Array.isArray(obj) ||
      typeof obj !== "object"
    ) {
      return true;
    }
    return Object.getOwnPropertyNames(obj).length === 0 ? true : false;
  }
}
