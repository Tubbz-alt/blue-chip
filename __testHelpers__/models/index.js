import Checklist from "./Checklist";
import Task from "./Task";
import Contact from "./Contact";
import PurchaseOrder from "./PurchaseOrder";
import PurchaseOrderContact from "./PurchaseOrderContact";
import Vendor from "./Vendor";
import VendorContact from "./VendorContact";
import Spec from "./Spec";
import SpecDetail from "./SpecDetail";
import SpecDetailCom from "./SpecDetailCom";

export {
  Checklist,
  Task,
  Contact,
  PurchaseOrder,
  PurchaseOrderContact,
  Vendor,
  VendorContact,
  Spec,
  SpecDetail,
  SpecDetailCom
};
