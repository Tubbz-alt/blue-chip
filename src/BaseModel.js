import pluralize from "pluralize";
import Query from "./Query";
import {lowerCaseFirst} from "./utils";

export default class BaseModel {
  static query(resources) {
    return new Query(
      this,
      this.pluralName(),
      resources,
      this.hasMany,
      this.belongsTo
    );
  }

  static pluralName() {
    return this.plural ? this.plural : lowerCaseFirst(pluralize(this.name));
  }

  static singularName() {
    return this.singular
      ? this.singular
      : lowerCaseFirst(pluralize(this.name, 1));
  }

  constructor(resources, attributes, hasMany = [], belongsTo = []) {
    Object.entries(attributes).forEach(([key, value]) => {
      this[key] = value;
    });

    if (hasMany.forEach) {
      hasMany.forEach(relationship =>
        this._buildHasManyQuery(this, resources, relationship)
      );
    }

    if (belongsTo.forEach) {
      belongsTo.forEach(relationship => {
        const relationshipKey = relationship.singularName();
        if (!this[relationshipKey]) {
          this[relationshipKey] = () => {
            const ParentClass = relationship;
            const ChildClass = this.constructor;
            return ParentClass.query(resources)
              .where({id: [this.id]})
              .toModels()[0];

            // Todo: belongsTo where related doesn't appear to work
            // return ParentClass.query(resources)
            //   .whereRelated(ChildClass, {id: this.id})
            //   .toModels()[0];
          };
        }
      });
    }
  }

  _filterResources(resource, resources, relationship, relationshipKey) {
    const currentResourceKey = resource.constructor.pluralName();

    const resourceClass = resource.constructor;
    const relationshipClass = relationship;
    return {
      ...resources,
      [currentResourceKey]: resources[currentResourceKey][resource.id],
      [relationshipKey]: relationshipClass
        .query(resources)
        .whereRelated(resourceClass, {
          id: resource.id
        }).currentResources
    };
  }

  _buildHasManyQuery(resource, resources, relationship) {
    const relationshipKey = relationship.pluralName();
    if (!resource[relationshipKey]) {
      resource[relationshipKey] = () => {
        const newResouces = resource._filterResources(
          resource,
          resources,
          relationship,
          relationshipKey
        );

        return new Query(
          relationship,
          relationshipKey,
          newResouces,
          relationship.hasMany,
          relationship.belongsTo
        );
      };
    }
  }
}
