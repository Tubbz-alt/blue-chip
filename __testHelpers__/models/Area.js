import BaseModel from "../../src/BaseModel";
import Spec from "./Spec";
import AreaType from "./AreaType";
import AreaRoom from "./AreaRoom";

export default class Area extends BaseModel {
  static get belongsTo() {
    return [Spec, AreaType];
  }
  static get hasMany(){
    return [AreaRoom];
  }
}
