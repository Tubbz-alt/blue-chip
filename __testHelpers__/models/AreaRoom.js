import BaseModel from "../../src/BaseModel";
import Area from "./Area";

export default class AreaRoom extends BaseModel {
  static get belongsTo(){
    return [Area]
  }
}
