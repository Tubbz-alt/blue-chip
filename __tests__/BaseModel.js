import BaseModel from "../src/BaseModel";
import Query from "../src/Query";

import resources from "../__testHelpers__/fixtrues/checklistsAndTasksNormalized";
import {Checklist, Task} from "../__testHelpers__/models";

describe("BaseModel", () => {
  test("query returns a Query", () => {
    expect(BaseModel.query(resources)).toBeInstanceOf(Query);
  });

  test("handle undefined resources", () => {
    expect(
      Checklist.query()
        .all()
        .toModels()
    ).toEqual([]);
    expect(
      Checklist.query(null)
        .all()
        .toModels()
    ).toEqual([]);
    expect(
      Checklist.query()
        .all()
        .toObjects()
    ).toEqual([]);
    expect(
      Checklist.query(null)
        .all()
        .toObjects()
    ).toEqual([]);
    expect(
      Checklist.query({})
        .all()
        .toObjects()
    ).toEqual([]);
    const a = {};
    expect(
      Checklist.query(a.b)
        .all()
        .toObjects()
    ).toEqual([]);
  });

  describe("models", () => {
    describe("all()", () => {
      test("Returns a list of models", () => {
        expect(
          Checklist.query(resources)
            .all()
            .toModels()
        ).toMatchSnapshot();
      });
    });

    describe("find()", () => {
      test("returns a single model", () => {
        const checklist = Checklist.query(resources).find(1);
        expect(checklist.id).toEqual(1);
        expect(checklist).toMatchSnapshot();
      });
    });

    describe("first()", () => {
      test("returns the first record in the store BY resource index", () => {
        const checklist = Checklist.query(resources).first();
        expect(checklist.id).toEqual(3);
        expect(checklist).toMatchSnapshot();
      });
    });

    describe("last()", () => {
      test("returns the last record in the store BY resource index", () => {
        const checklist = Checklist.query(resources).last();
        expect(checklist.id).toEqual(2);
        expect(checklist).toMatchSnapshot();
      });
    });

    describe("where()", () => {
      test("filters by name", () => {
        const checklist = Checklist.query(resources)
          .where({name: "Project Audit Rest"})
          .toModels()[0];
        expect(checklist.id).toEqual(2);
        expect(checklist).toMatchSnapshot();
      });

      test("filter by id", () => {
        const checklist = Checklist.query(resources)
          .where({id: 3})
          .toModels()[0];
        expect(checklist.id).toEqual(3);
        expect(checklist).toMatchSnapshot();
      });
    });

    describe("includes() hasMany", () => {
      test("check for hasMany includes", () => {
        const checklists = Checklist.query(resources)
          .all()
          .includes(["tasks"])
          .toModels();
        expect(checklists).toMatchSnapshot();
      });
    });

    describe("includes() belongsTo", () => {
      test("", () => {
        const tasks = Task.query(resources)
          .all()
          .includes(["checklist"])
          .toModels();
        expect(tasks).toMatchSnapshot();
      });
    });

    describe("hasMany()", () => {
      test("hasMany function", () => {
        const tasks = Checklist.query(resources)
          .find(1)
          .tasks()
          .toModels();
        expect(tasks).toMatchSnapshot();
      });
    });

    describe("belongsTo()", () => {
      test("", () => {
        const checklist = Task.query(resources)
          .find(1)
          .checklist();
        expect(checklist).toMatchSnapshot();
      });
    });
  });
  describe("objects", () => {
    describe("all()", () => {
      test("", () => {
        expect(
          Checklist.query(resources)
            .all()
            .toObjects()
        ).toMatchSnapshot();
      });
    });

    describe("find()", () => {
      test("", () => {
        const checklist = Checklist.query(resources).find(2);
        expect(checklist.id).toEqual(2);
        expect(checklist).toMatchSnapshot();
      });
    });

    describe("where()", () => {
      test("", () => {
        const checklist = Checklist.query(resources)
          .where({name: "Project Audit Rest"})
          .toObjects()[0];
        expect(checklist.id).toEqual(2);
        expect(checklist).toMatchSnapshot();
      });
    });

    describe("includes()", () => {
      test("", () => {
        const checklists = Checklist.query(resources)
          .all()
          .includes(["tasks"])
          .toObjects();
        expect(checklists).toMatchSnapshot();
      });
    });
  });
});
