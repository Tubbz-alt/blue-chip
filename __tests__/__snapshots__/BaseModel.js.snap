// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BaseModel models all() Returns a list of models 1`] = `
Array [
  Checklist {
    "id": 3,
    "name": "QA Checklist Rest",
    "tasks": [Function],
  },
  Checklist {
    "id": 1,
    "name": "Onboarding Rest",
    "tasks": [Function],
  },
  Checklist {
    "id": 2,
    "name": "Project Audit Rest",
    "tasks": [Function],
  },
]
`;

exports[`BaseModel models belongsTo()  1`] = `
Checklist {
  "id": 1,
  "name": "Onboarding Rest",
  "tasks": [Function],
}
`;

exports[`BaseModel models find() returns a single model 1`] = `
Checklist {
  "id": 1,
  "name": "Onboarding Rest",
  "tasks": [Function],
}
`;

exports[`BaseModel models first() returns the first record in the store BY resource index 1`] = `
Object {
  "attributes": Object {
    "name": "QA Checklist Rest",
  },
  "id": 3,
  "links": Object {
    "self": "http://example.com/checklists/3",
  },
  "relationships": Object {
    "tasks": Object {
      "data": Array [
        Object {
          "id": 5,
          "type": "tasks",
        },
        Object {
          "id": 6,
          "type": "tasks",
        },
      ],
    },
  },
  "type": "checklists",
}
`;

exports[`BaseModel models hasMany() hasMany 1`] = `
Array [
  Task {
    "checklist": [Function],
    "description": "Onboarding REST 1",
    "id": 1,
    "user": [Function],
  },
  Task {
    "checklist": [Function],
    "description": "Onboarding REST 2",
    "id": 2,
    "user": [Function],
  },
]
`;

exports[`BaseModel models includes() belongsTo  1`] = `
Array [
  Task {
    "checklist": Checklist {
      "id": 1,
      "name": "Onboarding Rest",
    },
    "description": "Onboarding REST 1",
    "id": 1,
    "user": [Function],
  },
  Task {
    "checklist": Checklist {
      "id": 1,
      "name": "Onboarding Rest",
    },
    "description": "Onboarding REST 2",
    "id": 2,
    "user": [Function],
  },
  Task {
    "checklist": Checklist {
      "id": 2,
      "name": "Project Audit Rest",
    },
    "description": "Project Audit Rest 3",
    "id": 3,
    "user": [Function],
  },
  Task {
    "checklist": Checklist {
      "id": 2,
      "name": "Project Audit Rest",
    },
    "description": "Project Audit Rest 4",
    "id": 4,
    "user": [Function],
  },
  Task {
    "checklist": Checklist {
      "id": 3,
      "name": "QA Checklist Rest",
    },
    "description": "QA Checklist Rest 5",
    "id": 5,
    "user": [Function],
  },
  Task {
    "checklist": Checklist {
      "id": 3,
      "name": "QA Checklist Rest",
    },
    "description": "QA Checklist Rest 6",
    "id": 6,
    "user": [Function],
  },
]
`;

exports[`BaseModel models includes() hasMany  1`] = `
Array [
  Checklist {
    "id": 3,
    "name": "QA Checklist Rest",
    "tasks": Array [
      Task {
        "description": "QA Checklist Rest 5",
        "id": 5,
      },
      Task {
        "description": "QA Checklist Rest 6",
        "id": 6,
      },
    ],
  },
  Checklist {
    "id": 1,
    "name": "Onboarding Rest",
    "tasks": Array [
      Task {
        "description": "Onboarding REST 1",
        "id": 1,
      },
      Task {
        "description": "Onboarding REST 2",
        "id": 2,
      },
    ],
  },
  Checklist {
    "id": 2,
    "name": "Project Audit Rest",
    "tasks": Array [
      Task {
        "description": "Project Audit Rest 3",
        "id": 3,
      },
      Task {
        "description": "Project Audit Rest 4",
        "id": 4,
      },
    ],
  },
]
`;

exports[`BaseModel models last() returns the last record in the store BY resource index 1`] = `
Object {
  "attributes": Object {
    "name": "Project Audit Rest",
  },
  "id": 2,
  "links": Object {
    "self": "http://example.com/checklists/2",
  },
  "relationships": Object {
    "tasks": Object {
      "data": Array [
        Object {
          "id": 3,
          "type": "tasks",
        },
        Object {
          "id": 4,
          "type": "tasks",
        },
      ],
    },
  },
  "type": "checklists",
}
`;

exports[`BaseModel models where() filter by id 1`] = `
Checklist {
  "id": 3,
  "name": "QA Checklist Rest",
  "tasks": [Function],
}
`;

exports[`BaseModel models where() filters by name 1`] = `
Checklist {
  "id": 2,
  "name": "Project Audit Rest",
  "tasks": [Function],
}
`;

exports[`BaseModel objects all()  1`] = `
Array [
  Object {
    "id": 3,
    "name": "QA Checklist Rest",
  },
  Object {
    "id": 1,
    "name": "Onboarding Rest",
  },
  Object {
    "id": 2,
    "name": "Project Audit Rest",
  },
]
`;

exports[`BaseModel objects find()  1`] = `
Checklist {
  "id": 2,
  "name": "Project Audit Rest",
  "tasks": [Function],
}
`;

exports[`BaseModel objects includes()  1`] = `
Array [
  Object {
    "id": 3,
    "name": "QA Checklist Rest",
    "tasks": Array [
      Object {
        "description": "QA Checklist Rest 5",
        "id": 5,
      },
      Object {
        "description": "QA Checklist Rest 6",
        "id": 6,
      },
    ],
  },
  Object {
    "id": 1,
    "name": "Onboarding Rest",
    "tasks": Array [
      Object {
        "description": "Onboarding REST 1",
        "id": 1,
      },
      Object {
        "description": "Onboarding REST 2",
        "id": 2,
      },
    ],
  },
  Object {
    "id": 2,
    "name": "Project Audit Rest",
    "tasks": Array [
      Object {
        "description": "Project Audit Rest 3",
        "id": 3,
      },
      Object {
        "description": "Project Audit Rest 4",
        "id": 4,
      },
    ],
  },
]
`;

exports[`BaseModel objects where()  1`] = `
Object {
  "id": 2,
  "name": "Project Audit Rest",
}
`;
